### ISSUE TITLE
Multiple in-flight requests for one key

### ISSUE THREAD

**Issue by pertu** [ben-manes/caffeine#625]
Demo project: https://github.com/pertu/caffeine-async-loading-in-flight-ignored (`mvn test` should work).

Test from the demo project (simplified):
```java
AsyncCacheLoader<Integer, Integer> loader =
      /* the future never completes = always "in flight" */
      (key, ignoredExecutor) -> new CompletableFuture<>();

/* I want this cache to reuse simultaneous/in-flight requests
   but never actually cache results */
var cacheInFlight = Caffeine.newBuilder()
      .expireAfterWrite(Duration.ZERO)
      // duration > 1ms makes the test pass:
      //.expireAfterWrite(Duration.ofMillis(1))
      .buildAsync(loader);

var future1 = cacheInFlight.get(1);
// this sleep() also makes the test pass:
//Thread.sleep(1);
var future2 = cacheInFlight.get(1);

// I expected these futures to be the same one future
// but this fails (sometimes):
assertSame(future1, future2);
```
See also: #297  
Edit: tested version 3.0.4 and 2.9.2, result is the same  
Edit 2: the test above is timing-dependent and doesn't always fail
actual test in demo project runs the code above in loop and usually fails on the second iteration on my machine

**Comment by ben-manes** [comment]
I won't get a chance to look at this for a few hours, but fwiw you can use `Caffeine.ticker(Ticker)` to control time in tests, e.g. by using Guava's FakeTicker. That might help you debug it.

**Comment by ben-manes** [comment]
The mistake is that the expiration check is performed prior to an async check and it does not handle if the entry is still in-flight. I think this occurs in other read paths, like `getIfPresent`, as all check for async only with respect to setting the new timestamps.

https://github.com/ben-manes/caffeine/blob/f924b5c07c4defb148101cd1c17cc033df0acfb1/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java#L2418-L2443

Previously `hasExpired` used to check if an in-flight entry, but that seems to have been lost in v2.6.2 when fixing an [overflow](https://github.com/ben-manes/caffeine/commit/111925379d39a82eff0b942891cac674dd157f3c) scenario. The rational might have been that if the loading time exceeds the duration then it might be stale, but I am not certain without reading the ticket and code changes more thoroughly. Offhand I'd agree with you and the previous logic did this, so we'll need to understand if this is a regression or expected behavior.

https://github.com/ben-manes/caffeine/blob/6be13d692041106e51656dc524c52a2bfd68a76e/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java#L727-L735

**Comment by ben-manes** [comment]
I believe this was accidentally removed when debugging to simulate the overflow issue in #217. The commit's description, changes, and discussions don't appear relevant to this snippet. The test suite passes when it is restored.

I will add the appropriate tests to catch this from regressing in the future. I hope to work on it this weekend provide you a release by Monday for both 2.x and 3.x. In the meantime or if that is delayed, you can use the v3.0.4-SNAPSHOT jar on Sonatype's repository or jitpack the `v3.dev` branch. Thanks for reporting this issue and providing a test case.

**Comment by pertu** [comment]
It's a pleasure to report issues when they are answered. Thanks for the quick response and analysis.
I guess there is no urgency, as this issue was out there for quite some time already and no one reported it.

**Comment by koldat** [comment]
I am not sure if it is related to this problem, but what is happening to me is that sometimes my loader is never called again, because Node still think it is running. Debugger says that future is complete (it also returns immediately when got from cache)  as well as my job status of the async loader. But Node in the cache still has writeTime set to large value (200 years ahead - com.github.benmanes.caffeine.cache.Async.ASYNC_EXPIRY) which prevents expiration. That future time from what I understand is marker that task is still running.

Do you think it is same issue?

**Comment by ben-manes** [comment]
That would indicate that the callback is never invoked when the future completed. In this issue the future was prematurely replaced with a new execution, but when that replacement completed it's handler should successfully update the `writeTime`. The handler calls ` cache().replace(key, castedFuture, castedFuture);` in the code below to update the metadata. If you can provide a reproducer then I can investigate that in a new issue.

https://github.com/ben-manes/caffeine/blob/86991daa56e83c4d69ad977639644a94c44ddec3/caffeine/src/main/java/com/github/benmanes/caffeine/cache/LocalAsyncCache.java#L189-L217

**Comment by koldat** [comment]
OK I have found the problem. It actually works fine, but thread that loads asynchronously is then used to "complete" future and notify. In my case I convert future to Mono in reactor and it steals thread to execute requests. This go back to the cache again and cycles (I do retry there). That as a result is actually working, but on incorrect place and cannot get back. Caffeine is then still thinking future is working on result even if it finished and is still in notify. Sorry for confusion.

**Comment by ben-manes** [comment]
I did not get to working on this today, where the intent was to backport to v2 and release. I will see what I can do over the week, but given the holidays and an ongoing investigations for the root cause of f143764, the goal for a release is delayed until the next weekend. The jitpack or snapshot jars pass our CI testing and should be stable if you'd prefer something earlier.

@koldat it sounds like you ran into the surprising fact that CompletableFuture's ordering of dependent actions is in LIFO order for efficiency, whereas one might naturally expect FIFO execution. Unfortunately we cannot work around this quirk without worse effects. If we instead returned the `whenComplete` future then the cache returns a different instance than was obtained by the loader and stored, but only on the creation's call. If we added a `whenComplete` variant within the computation, to at least return what is stored, then a direct execution could fail due to a recursive write to `ConcurrentHashMap`. Any logic that depended on the source future instance could not write concurrent logic using `AsyncCache.asMap().replace(key, oldValue, newValue)` or call a mutation method like `future.cancel()`. Therefore the best is to be consistent and treat the lack of specificity for dependent action ordering as not something that we can resolve.
